CORS-3923, CORS-3927: Support confidential cluster installation on SEV-SNP and TDX nodes on GCP

bgartzi opened the pull request:
This patch series aims to support installing a cluster on AMD SEV SNP or TDX confidential nodes on GCP.
Previously, only AMD SEV nodes were supported, which were configured through the `confidentialCompute: Enabled` configuration flag.
Now that GCP supports specifying the confidential instance type, we are letting users specify which node type (SEV/SEV-SNP/TDX) they would like to deploy the cluster on.
This can be done by using any of the new available values in `condidentialCompute` such as `AMDEncryptedVirtualization` (similar to `Enabled`), `AMDEncryptedVirtualizationNestedPaging` (AMD SEV-SNP) or `IntelTrustedDomainExtensions` (Intel TDX).

This series depend on the following patches:
- https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/1410
- https://github.com/openshift/api/pull/2165
- https://github.com/openshift/machine-api-operator/pull/1326
- https://github.com/openshift/machine-api-provider-gcp/pull/110

Which have been merged now and and vendored in this patch.

A new cluster-api-provider-gcp release containing the changes I submitted isn't available yet. I created the `data/data/cluster-api/gcp-infrastructure-components.yaml` myself. I'm not sure if that's a valid approach, or if we should rather wait until a new version is out.

barbacbd commented:
/uncc

jianli-wei commented:
@bgartzi Suggestions on the explain command outputs. 

(1) the current outputs:
```
$ ./openshift-install explain installconfig.compute.platform.gcp.confidentialCompute
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <string>
  Valid Values: "","Enabled","Disabled","AMDEncryptedVirtualization","AMDEncryptedVirtualizationNestedPaging","IntelTrustedDomainExtensions"
  confidentialCompute is an optional field defining whether the instance should have confidential compute enabled or not, and the confidential computing technology of choice.
Allowed values are omitted, Disabled, Enabled, AMDEncryptedVirtualization, AMDEncryptedVirtualizationNestedPaging, and IntelTrustedDomainExtensions
When set to Disabled, the machine will not be configured to be a confidential computing instance.
When set to Enabled, the machine will be configured as a confidential computing instance with no preference on the confidential compute policy used. In this mode, the platform chooses a default that is subject to change over time. Currently, the default is to use AMD Secure Encrypted Virtualization.
When set to AMDEncryptedVirtualization, the machine will be configured as a confidential computing instance with AMD Secure Encrypted Virtualization (AMD SEV) as the confidential computing technology.
When set to AMDEncryptedVirtualizationNestedPaging, the machine will be configured as a confidential computing instance with AMD Secure Encrypted Virtualization Secure Nested Paging (AMD SEV-SNP) as the confidential computing technology.
When set to IntelTrustedDomainExtensions, the machine will be configured as a confidential computing instance with Intel Trusted Domain Extensions (Intel TDX) as the confidential computing technology.
If any value other than Disabled is set the selected machine type must support that specific confidential computing technology. The machine series supporting confidential computing technologies can be checked at https://cloud.google.com/confidential-computing/confidential-vm/docs/supported-configurations#all-confidential-vm-instances
Currently, AMDEncryptedVirtualization is supported in c2d, n2d, and c3d machines.
AMDEncryptedVirtualizationNestedPaging is supported in n2d machines.
IntelTrustedDomainExtensions is supported in c3 machines.
If any value other than Disabled is set, the selected region must support that specific confidential computing technology. The list of regions supporting confidential computing technologies can be checked at https://cloud.google.com/confidential-computing/confidential-vm/docs/supported-configurations#supported-zones
If any value other than Disabled is set onHostMaintenance is required to be set to "Terminate".
If omitted, the platform chooses a default, which is subject to change over time, currently that default is Disabled.

$ 
```
(2) one sample outputs:
```
$ openshift-install explain installconfig.platform.gcp.userTags
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <[]object>
  userTags has additional keys and values that the installer will add as
tags to all resources that it creates on GCP. Resources created by the
cluster itself may not include these tags. Tag key and tag value should
be the shortnames of the tag key and tag value resource.

FIELDS:
    key <string> -required-
      key is the key part of the tag. A tag key can have a maximum of 63 characters and
cannot be empty. Tag key must begin and end with an alphanumeric character, and
must contain only uppercase, lowercase alphanumeric characters, and the following
special characters `._-`.

    parentID <string> -required-
      parentID is the ID of the hierarchical resource where the tags are defined,
e.g. at the Organization or the Project level. To find the Organization ID or Project ID refer to the following pages:
https://cloud.google.com/resource-manager/docs/creating-managing-organization#retrieving_your_organization_id,
https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects.
An OrganizationID must consist of decimal numbers, and cannot have leading zeroes.
A ProjectID must be 6 to 30 characters in length, can only contain lowercase letters,
numbers, and hyphens, and must start with a letter, and cannot end with a hyphen.

    value <string> -required-
      value is the value part of the tag. A tag value can have a maximum of 63 characters
and cannot be empty. Tag value must begin and end with an alphanumeric character, and
must contain only uppercase, lowercase alphanumeric characters, and the following
special characters `_-.@%=+:,*#&(){}[]` and spaces.

$ 
```
(3) The suggestions:
- Make the max length of each line being 85~90 (as the sample outputs).
- Remove the statement `Allowed values are omitted, Disabled, Enabled, AMDEncryptedVirtualization, AMDEncryptedVirtualizationNestedPaging, and IntelTrustedDomainExtensions`, as the "Valid Values: " line had told it. 
- Remove the statement `If any value other than Disabled is set, the selected region must support that specific confidential computing technology. The list of regions supporting confidential computing technologies can be checked at https://cloud.google.com/confidential-computing/confidential-vm/docs/supported-configurations#supported-zones`.
- Remove the statement `If omitted, the platform chooses a default, which is subject to change over time, currently that default is Disabled.`.
- Replace statements like `When set to Disabled` with "  With Disabled" (please note the 2 leading spaces).
- Replace statements like `the machine will (not )?be configured as a confidential computing instance` with "confidential computing is enabled|disabled".
- Replace statement `When set to Enabled, the machine will be configured as a confidential computing instance with no preference on the confidential compute policy used. In this mode, the platform chooses a default that is subject to change over time. Currently, the default is to use AMD Secure Encrypted Virtualization.` with "  With Enabled, confidential computing is enabled, but without preferred Confidential Computing technology. The platform chooses a default, i.e. AMD SEV-SNP, which is subject to change over time. "
- Replace statement `If any value other than Disabled is set the selected machine type must support that specific confidential computing technology. ` with "If any value other than Disabled is set, the machine type, which supports the selected Confidential Computing technology, must be specified. "
- In `If any value other than Disabled is set the selected machine type must support that specific confidential computing technology. The machine series supporting confidential computing technologies can be checked at https://cloud.google.com/confidential-computing/confidential-vm/docs/supported-configurations#all-confidential-vm-instances`, please replace the link with "https://cloud.google.com/confidential-computing/confidential-vm/docs/supported-configurations#machine-type-cpu-zone".
- With the above link, suggest to remove lines `Currently, AMDEncryptedVirtualization is supported in c2d, n2d, and c3d machines.
AMDEncryptedVirtualizationNestedPaging is supported in n2d machines.
IntelTrustedDomainExtensions is supported in c3 machines.
`

sadasu commented:
/retest-required

jianli-wei commented:
> I just addressed your concerns in the last force-pushed version. Could you confir @jianli-wei?

@bgartzi Thanks for the quick action! The updates looks good to me, just 2 minor suggestions. 

(1) The current outputs:
```
$ ./openshift-install explain installconfig.platform.gcp.defaultMachinePlatform.confidentialCompute
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <string>
  Default: "Disabled"
  Valid Values: "","Enabled","Disabled","AMDEncryptedVirtualization","AMDEncryptedVirtualizationNestedPaging","IntelTrustedDomainExtensions"
  confidentialCompute is an optional field defining whether the instance should have
confidential compute enabled or not, and the Confidential Computing technology of choice.
 With Disabled, Confidential Computing is disabled.
 With Enabled, Confidential Computing is enabled with no preference on the
Confidential Compute policy. The platform chooses a default i.e. AMD SEV,
which is subject to change over time.
 With AMDEncryptedVirtualization, Confidential Computing is enabled with
AMD Secure Encrypted Virtualization (AMD SEV).
 With AMDEncryptedVirtualizationNestedPaging, Confidential Computing is
enabled with AMD Secure Encrypted Virtualization Secure Nested Paging
(AMD SEV-SNP).
 With IntelTrustedDomainExtensions, Confidential Computing is enabled with
Intel Trusted Domain Extensions (Intel TDX).
If any value other than Disabled is set, a machine type and region that supports
Confidential Computing must be specified. Machine series and regions supporting
Confidential Computing technologies can be checked at
https://cloud.google.com/confidential-computing/confidential-vm/docs/supported-configurations#machine-type-cpu-zone
If any value other than Disabled is set onHostMaintenance is required to be set
to "Terminate".

$ 
```

(2) Minor suggestions:
- Replace "confidential compute" with "Confidential Computing", and replace "Confidential Compute policy" with "Confidential Computing technologies".
- Insert 4 leading spaces before the statements like `With {Disabled|Enabled|AMDEncryptedVirtualization|AMDEncryptedVirtualizationNestedPaging|IntelTrustedDomainExtensions}, `, and the statements `If any value other than Disabled is set`.
- Insert a comma after the word "set", in `If any value other than Disabled is set onHostMaintenance is required...`.

FYI with the above suggestions, the outputs should look like, 
```
$ ./openshift-install explain installconfig.platform.gcp.defaultMachinePlatform.confidentialCompute
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <string>
  Default: "Disabled"
  Valid Values: "","Enabled","Disabled","AMDEncryptedVirtualization","AMDEncryptedVirtualizationNestedPaging","IntelTrustedDomainExtensions"
  confidentialCompute is an optional field defining whether the instance should have
Confidential Computing enabled or not, and the Confidential Computing technology of choice.
    With Disabled, Confidential Computing is disabled.
    With Enabled, Confidential Computing is enabled with no preference on the
Confidential Computing technologies. The platform chooses a default i.e. AMD SEV,
which is subject to change over time.
    With AMDEncryptedVirtualization, Confidential Computing is enabled with
AMD Secure Encrypted Virtualization (AMD SEV).
    With AMDEncryptedVirtualizationNestedPaging, Confidential Computing is
enabled with AMD Secure Encrypted Virtualization Secure Nested Paging
(AMD SEV-SNP).
    With IntelTrustedDomainExtensions, Confidential Computing is enabled with
Intel Trusted Domain Extensions (Intel TDX).
    If any value other than Disabled is set, a machine type and region that supports
Confidential Computing must be specified. Machine series and regions supporting
Confidential Computing technologies can be checked at
https://cloud.google.com/confidential-computing/confidential-vm/docs/supported-configurations#machine-type-cpu-zone
    If any value other than Disabled is set, onHostMaintenance is required to be set
to "Terminate".

$

patrickdillon commented:
/approve

bgartzi commented:
Hey @jianli-wei, @patrickdillon, thanks for the review and testing. I just pushed a new version. Comparing to the previous one:
- I rebased the branch on latest `main` due to conflicts.
- I dropped the commit bumping `openshift/api` as the commits I needed are included in the latest vendored version.
- I addressed `explain`'s output style concerns.
- I added a validation step that checks that the configured `confidentialCompute` matches the confidential computing technology supported by the configured instance type, as requested in https://issues.redhat.com/browse/CORS-3924.

bgartzi commented:
Fixing some linter an gofmt issues that the ci found...

bgartzi commented:
I just dropped the commit bumping the cluster-api-provider-gcp version. The changes I needed were already included by #9528.

patrickdillon commented:
/lgtm

jianli-wei commented:
/label qe-approved
